Refactor cluster manager and connection manager to use otp behaviors.

lordnull opened the pull request:
Splits out responsibilities a bit and uses otp behaviors (mainly gen_fsm) to help code-readability and introspection of running systems.

Riak_test is basho/riak_test/pull/684 .

kellymclaughlin committed:
Convert riak_core_cluster_conn to use the OTP gen_fsm behavior

Convert riak_core_cluster_conn to use the OTP gen_fsm behavior. The
changes affect the internal operation of the module, but the API is
unchanged. Also add an eqc property to exercise the fsm state
transitions.

macintux committed:
Fixes to make edoc happy

macintux committed:
More superficial cleanup: consistency, typos, fix module comments that were unintentionally attributed to the first function

macintux committed:
Fix dialyzer warnings after line adjustments

macintux committed:
Moar monospace

macintux committed:
Remember how to monospace

jtuple committed:
Extend AAE fullsync to use a pipelined exchange

The current AAE fullsync exchange was based on the exchange FSM used by
Riak for local cluster AAE. However, this was a bad design decision as
local AAE was never designed to tolerate high latency links.

In the worst case, this design can require millions of sequential
roundtrip messages. This is not an issue for a fast LAN, but is
impractical for WANs with 10-200ms RTTs.

This commit changes the AAE fullsync exchange to use a pipelined approach.
Rather than performing a synchronous request for each bucket, the exchange
determines all differences for a given level of the AAE tree and then
requests the data for all differing buckets upfront.

Ideally, this would supported by a new streaming protocol. However to
retain protocol compatibility the pipelined approach sends all the
known bucket requests upfront and relies upon TCP ordering to deliver
the responses in the order requested.

Note: this commit relies upon changes made to riak_core and riak_kv to
extend the AAE subsystem to support the new level-by-level exchange that
is utilized by this new fullsync approach.

jtuple committed:
Refactor AAE FS source state transitions for clarity

This commit slightly refactors the riak_repl_aae_source code to
make state transitions more obvious.

This change also eases the porting of commits from a separate feature
branch that implemented a slightly different approach to improved AAE
syncing.

jtuple committed:
Add buffer-then-send mode for AAE fullsync repair

This commit merges in the separately prototyped buffer-then-send
approach to directly sending key differences. The mode used is
determined by the `fullsync_direct_mode` application variable,
which defaults to `inline`.

* inline
  differences selected for direct repair are sent as they are encountered
  during an AAE exchange.

* buffered
  differences selected for direct repair are buffered in memory during an
  exchange. After the exchange, the buffer is sorted and keys are then sent
  in order.

krestenkrab committed:
Only do one backend fold for each vnode, not N

Previous code created a bloom filter for each
IndexN, populating it with diffs, and then did
a backend fold for that vnode sending objects
along the way.

The code now

1. creates the bloom filter in init()
2. runs key exchange for each IndexN
3. then does a vnode fold to find objects in the
   bloom filter.

krestenkrab committed:
Use N=1 for reading values for replication

krestenkrab committed:
Adaptable sync strategy: get or fold w/threshold

This change introduces a threshold for how
to process fullsync diffs.  A small number of
changed KVs (currently fixed at 1000 per vnode)
are read using GET; above that limit, objects
are read using a backend fold.

For small number og changed KVs, this avoids
doing a full scan of the backend data.

jtuple committed:
Make fullsync direct send threshold configurable

jaredmorrow committed:
Set rebar.config deps back to 2.0 branches

lordnull committed:
Converted riak_core_connection to a gen_fsm.

lordnull committed:
Fixed riak_core_connect so tests pass.

lordnull committed:
Fixed stop returns and active-ness of socket.

lordnull committed:
Fixed connection to return useful messages instead of link and crash

This will remidied to be more idiomatic erlang later, but for now
this seems to be the intent of the test.

lordnull committed:
Extracted connection logic from service_mgr.

Also made some tests able to activate lager, and added some debug logging.

lordnull committed:
Extracted cluster service code to its own module.

lordnull committed:
Extracted service connection logic to it's own module.

Commit:
Added a worker pool for fullsync sinks.

Although keylisting has a pool for issuing gets, both the
AAE and the Keylist strategies serialize writing objects
from the sink process.  This adds a pool with a default size
of 100 shared across all fullsync sinks that make the writes async.

No difference in safety properties, neither version checks
the response from the put FSM.

jtuple committed:
Fix deadlock between AAE fullsync sink/source

A deadlock in AAE fullsync was identified during testing.

The deadlock occurs when both the AAE fullsync source and sink
processes block on TCP send due to TCP backpressure. Since both
processes are blocked, neither can perform a receive to unblock
the other.

This commit addresses the issue by making the AAE sink use a helper
process for sending, therefore ensuring the sink never blocks on a
send. The sink will therefore always get around to receiving data at
some point, unblocking the source if needed.

NOTE: the SSL transport wraps the underlying socket in a gen_server, and
all socket functions route through this server. As such, the SSL transport
is still vulnerable to this deadlock. Fixing this requires either changing
the AAE protocol or patching the OTP SSL implementation. One of the two is
planned future work.

jtuple committed:
Remove non-pipelined AAE fullsync mode

lordnull committed:
Gernalized transport handling.

jtuple committed:
Change AAE FS source to use direct vnode gets

When performing direct sends, the AAE fullsync source retrieves
individual objects in order to send them. Previously, this was
accomplished using normal riak_client gets using the n_val=1
option. However, there are no guarantees about which replica
the object would come from using such an approach.

This commit changes the AAE fullsync source to instead retrieve
objects directly from the vnode performing the fullsync. This
ensures that the object sent aligns with the AAE hash used to
determine differences. This approach may also have better
performance.

Commit:
Changed fullsync_direct to fullsync_direct_limit, stopped "Gathering source data" msg.

Minor changes for user experience, clarified with _limit to permit a future
_percentage version of fullsync_direct.

Lowered the "Gathering source data" message displayed every minute to a
debug message.

jtuple committed:
Merge pull request #618 from branch 'feature/fssink-pool-2.0'

Reviewed-by: jtuple

jtuple committed:
Merge pull request #619 from branch 'feature/pipelined-aae-fullsync-ux-2.0'

Reviewed-by: jtuple

borshop committed:
Merge pull request #620 from basho/feature/pipelined-aae-fullsync-2.0

Improved AAE fullsync integration/2.0 pull request

Reviewed-by: jonmeredith

lordnull committed:
Make cluster conn sup kids transient.

This means the kids, on exit of normal/shutdown, do not yell and scream
for no good reason.

lordnull committed:
Fixes so replicaiton tests in r_t pass.

lordnull committed:
Fixed mis-handling of closed connection.

Because a closed connection is likely a remote error, it is better to
exit non-normally and have a supervisor restart than to simply silently
exit.

krestenkrab commented on src/riak_core_service_conn.erl:
```diff
@@ -0,0 +1,236 @@
+%% ===============
+
+handle_info({_TransTag, Socket, Data}, wait_for_hello, State = #state{socket = Socket}) when is_binary(Data) ->
+    case binary_to_term(Data) of
```
Urgh! We need to play it more safe here, because we're decoding random data coming off the internet.  I suggest limiting the size of `Data`, and also using `binary_to_term(Data, [safe])`, so that someone posing as a remote cluster cannot inject random new atoms into this VM.  Remember this all happens _before_ SSL upgrade.

It's a little tricky, because we want to be able to handle the case when they have capabilities we don't know of, i.e. atoms that are not necessarily present in this VM.  So maybe just limiting `byte_size(Data)` to something like a few K should suffice...

lordnull replied:
I agree that this is bad thing™ and needs to be fixed. Both sides use 'term_to_binary', so I'll need to inspect the code closely to ensure that we never send an atom that the remote side may not already have. My particular concern is the capabilities exchange step. If that's the case, I'm not sure anything short of a complete revamp of the handshake protocol would suffice.

krestenkrab commented on src/riak_core_service_conn.erl:
```diff
@@ -0,0 +1,236 @@
+    end;
+
+handle_info({_TransTag, Socket, Data}, wait_for_protocol_versions, State = #state{socket = Socket}) when is_binary(Data) ->
+    case binary_to_term(Data) of
```
Same applies here regarding `binary_to_term(..., [safe])` and/or size limiting. If we're not on SSL, then this might well be data coming from a random client sitting somewhere on the net.

krestenkrab commented on src/riak_repl_console.erl:
```diff
@@ -321,7 +320,8 @@ showClusterConn({Remote,Pid}) ->
             CAddr = choose_best_addr(Remote, ClientAddr),
             io:format("~-20s ~-20s ~-15s ~p (via ~s)~n",
                       [ConnName, Name, PidStr,IPs, CAddr])
-    after 2 ->
```
From reading `riak_core_cluster_conn.erl`, it looks like the status could also be `initiating_connection`, `waiting_for_cluster_name`, and `waiting_for_cluster_members`.  Am I missing something?

krestenkrab commented:
All in all, minor things that need fixing, except for the indiscriminate decoding of `binary_to_term` on the open internet.

lordnull committed:
Added module documentation to cluster_conn.

Shamelessly taken from Kresten's documentation for the module in his
comment for a previous commit. Checked for accuracy and minor typos fixed.

lordnull committed:
Logging unmatched message in handle_info

lordnull committed:
Fixed tranpsort handle_infos to used Using Transport:messages().

This means it's easier to differentiate between transport data and
transport errors, and avoids hard-coding for tcp/ssl or other transports.

lordnull committed:
Fixed clust conn status call.

Console had vestibles of old status call, and did not handle all the
states for the cluster conn. The cluster conn did not reply with all the
info the console wanted. Fixed both.

lordnull committed:
Removed a controversial 'if' statement.

lordnull committed:
Added cluster serv checking of address info.

lordnull committed:
Added module documentation.

Taken from a code review comment by Kresten, and edited only for formatting.

lordnull committed:
Using the ranch tranport messages/0 api to better filter.

Instead of using the dodgey strategy of checking 'is_binary' or 'is_atom'
for socket messages, pull the message tag atoms from the ranch transport
and check against those. This should be forwards compatible even if new
transports beyond ssl and tcp are used.

krestenkrab committed:
Handle {pid(), poll_cluster} from cluster manager

krestenkrab committed:
Fix state bug in cluster conn update

krestenkrab commented:
After running (on code derived from this in branch `enhance/krab/cluster-realtime-rebalance`), we ran into two problems fixed with the following commits: 2c1744256c59e3804d50c1022beda4b3c4365aca and f41234feab6d202b355d3435d4488705b722dc82.

lordnull commented:
I've cherry-picked those two commits in on my local machine, though I'm not sure how that will effect your branch; whether you want to rebase it or keep the commits there (and me ditching the cherry-pick).

lordnull committed:
Fixed eunit tests to conform to changes.

krestenkrab commented:
No problem, @rsltrifork will work it out.

jonmeredith commented:
Looks like unit tests passed, but were marked as failure on the builder because it had to kill gold rush.

```
=======================================================
  All 189 tests passed.
Cover analysis: /usr/home/buildbot/masters/riak/test-riak_repl/build/.eunit/index.html
{"Kernel pid terminated",application_controller,"{application_terminated,goldrush,killed}"}

Crash dump was written to: erl_crash.dump
Kernel pid terminated (application_controller) ({application_terminated,goldrush,killed})
```

It also gave two dialyzer errors

```
==> build (dialyzer)
riak_core_connection.erl:171: The call ranch_tcp:send(Socket::port(),Hello::binary()) breaks the contract (inet:socket(),iolist()) -> 'ok' | {'error',atom()}
riak_core_connection.erl:263: Expression produces a value of type 'ok' | {'error',_}, but this value is unmatched
Unknown types:
  proplist:proplist/0
make: *** [dialyzer] Error 1
```

Can you investigate please @lordnull so we can merge this.

lordnull committed:
Fixes to allow dialyzer to pass.

Fixed a line number, a return value check, and a typoed type.

jonmeredith commented:
Passes Dialyzer now as of http://buildbot.bos1/builders/test-riak_repl/builds/1254, checking why GoldRush has problems on eunit shutdown.

jonmeredith commented:
@borshop: retry

Commit:
Increased timeout on riak_repl_leader test to try and fix skips.

lordnull committed:
Fixed lager startup to ease rebar eunit cleanup.

These changes were spured by galdrush exits killing the vm used to run
tests. Investigation showed rebar by default tries to clean up the vm state
after tests are run, and part of that is killing processes that were not
running before. Of the processes that were killed, the goldrush application
pid was one of them.

Goldrush is started by lager when using lager:start/0. lager starts itself
and it's dependency apps as permanent apps, meaning if they die they take
the kernel application manager with it, which takes the kernel pid, which
takes init, which takes the vm. By changing from lager:start/0 to
application:ensure_all_started/1, the started applications are temporary.
This means if they die, no-one actually cares. This also means lager
cleanup of the apps is done by stopping all started apps rather than just
lager.

lordnull commented:
Mistargeted pr, use #631 instead.
